### ISSUE TITLE
OMA-DM-LightweightM2M-2016-0119-CR_firmware_update_state_machine

### ISSUE THREAD

**Issue by boaks** [OpenMobileAlliance/OMA_LwM2M_for_Developers#136]
Would it be possible to add "Update fails [Res == 8]" to the arrow from "UPDATING" to "IDLE" to allow a "BootLoader" to delete a broken image? (Just if that broken image could only be detected by the bootloader itself?)

**Comment by boaks** [comment]
And I'm missing Res == 5 on an arrow to IDLE.
(May be added to the arrow from "DOWNLOADING" to "IDLE")

**Comment by hannestschofenig** [comment]
Good feedback! 

Res == 5 means 'Integrity check failure for new downloaded package.' This means that we would have to draw a transition from downloaded to idle. 

Creating a transition from Updating to Idle for the case where the firmware update fails also makes sense to me. 

Here is the new state machine: 

![statemachine-firmwareupdate_5sep2016](https://cloud.githubusercontent.com/assets/1016643/18252712/24c368aa-7390-11e6-830e-b6dbdb47075d.png)

**Comment by boaks** [comment]
In my opinion this may be suboptimal!
When "DOWNLOADED" is reported, someone may trigger the "Execute".

So my view, just before the CR was:
"DOWNLOADING": wait for URI validation, actual downloading, and for image validation. but mainly wait ...
"DOWNLOADED": OK to execute

**Comment by boaks** [comment]
But cool: the arrow from UPDATING to IDLE is added!

**Comment by hannestschofenig** [comment]
After giving it a second thought I agree that it is better to have a transition from UPDATING to IDLE for Res==5 (instead of having it of DOWNLOADED to IDLE).

![statemachine-firmwareupdate_5sep2016](https://cloud.githubusercontent.com/assets/1016643/18253157/bc60e690-7392-11e6-91d2-6a7b11c18561.png)

**Comment by boaks** [comment]
Ok, but sometimes the LWM2M application will also be able to do a integrity check.
So, adding that to "DOWNLOADING"  -> "IDLE" would also be nice :-).

**Comment by hannestschofenig** [comment]
I agree that it would be nice to do that while downloading but at the moment there is nothing in the LWM2M spec that would allow you do that. So, it would have to come through some other currently unspecified mechanism.

**Comment by hannestschofenig** [comment]
I added the requested transition since it might be useful in the future.

![statemachine-firmwareupdate_5sep2016](https://cloud.githubusercontent.com/assets/1016643/18254887/0b39ef22-73a3-11e6-839b-887c1438ca04.png)

**Comment by boaks** [comment]
> ... it would be nice to do that while downloading but at the moment there is nothing in the LWM2M spec that would allow you do that. 


Do you mean, while the transfer is ongoing?
So just rename the "DOWNLOADING" in "PREPARING", containing "Resolving the URI, do the actual transfer, and do the validation". Also just rename "DOWNLOADED" in "PREPARED" with the meaning, that the device is willing to accept a execute. This wording would also not require the introduction of e a new state. 

But even with the "DOWNLOADING", I don't see  the point!
Assuming the package is transfered using a LWM2M write to /5/0/0 (package), there is no issue not reporting a STATE/RESULT change, if a observe was already established. If no observe was established, then, depending on the actual technic, the transfer may be aborted (e.g. blockwise CoAP using other result code than CONTINUE) or the STATE/RESULT must be reported afterward (which would be the natural case for integrity check). And using /5/0/1 (package URL) the actual transfer is "out of scope of LWM2M"

**Comment by hannestschofenig** [comment]
The states have been named according to what is currently used in the LWM2M specification. So, I prefer not to change the names. 

Note that there are many ways to create a state machine and the currently described state machine is not normative, it is an example to better illustrate the process. I noted this in the text. 

What is normative is the need to expose the current system status via the state/res resource. I will have to explain this better in the text.

**Comment by boaks** [comment]
> What is normative is the need to expose the current system status via the state/res resource. I will have to explain this better in the text. 

OK, I thought, a new value "RESOLVE_URI" for STATE /5/0/3 should be introduced. 
But having a "closer" look, you didn't attach a "state" to "RESOLVE_URI".

**Comment by hannestschofenig** [comment]
Do you think that the state machine is clear enough to produce a new CR based on it?

**Comment by boaks** [comment]
I would add to "RESOLVE_URI" that the "state" may be 0 or 1, depending on the implementation.

And the removed "Firmware Update Failed" from UPDATING to DONWLOADED should, in my opinion also be added again. The reason for that maybe, that the bootloader detects a temporary situation, where the UPDATE is not possible at that time but may be later.

**Comment by boaks** [comment]
> Do you think that the state machine is clear enough to produce a new CR based on it?

With the above, I think so.

**Comment by hannestschofenig** [comment]
The State cannot be "1" at RESOLVE_URI since the firmware download only starts in the DOWNLOADING state. 

What is the meaning of a transition from UPDATING to DOWNLOADED under the condition the "firmware update failed"? How would an implementation get out of this state again? If we add that transition we should also provide a way to go from DOWNLOADED to IDLE since otherwise we are stuck.

**Comment by hannestschofenig** [comment]
I believe I have found the answer to my question in the spec ;-)

"Writing an empty string to Package Resource or to Package URI Resource, resets the Firmware Update State Machine: the State Resource value is set to Idle and the Update Result Resource value is set to 0."

So, there needs to be a transition from DOWNLOADED to IDLE when an empty string is written to the Packet resource or to the Package URI resource.

I changed the state machine slightly. 

![statemachine-firmwareupdate_6sep2016](https://cloud.githubusercontent.com/assets/1016643/18272364/c038d6a6-7437-11e6-9251-a2c7743cbeee.png)

**Comment by boaks** [comment]
> The State cannot be "1" at RESOLVE_URI since the firmware download only starts in the DOWNLOADING state.

This means, if, e.g. a DNS lookup takes some time, the device will not react on writing "/5/0/1" and a server, who expect a fast transition to DOWNLOADING will retry/rewrite "/5/0/1".

Therefore, the meaning of DOWNLOADING is in my opinion "preapring", the meaning of "DOWNLOADED" is "prepared". This doesn't change the Firmware Object (the state values stay), it just explains better, what exactly is expected.

**Comment by hannestschofenig** [comment]
Based on discussions during today's conference call I have simplified the state machine drawing. 

![statemachine-20sep2016](https://cloud.githubusercontent.com/assets/1016643/18678076/2bba73f2-7f5b-11e6-955e-abd27054dd45.png)

**Comment by boaks** [comment]
Hello Hannes,

looks great!
One question left for me:
A device, which detects an error (e.g. an invalid URL (7)),
MUST it first set the (state 1, result 0) and then (state 0, result 7)?
Or is it valid to report (state 0, result 7) immediately?

I would go for the second one, which would add the “error arrows” from IDLE to itself.
If so, a Note to clear a previous result before starting a new download may be added.

The first one, with MUST (state 1), will raise the question, how long it should stay
in 1 and if a LWM2M server can rely on that. I don’t think, it would be a good solution
to specify that the server can rely on get aware of state 1, because this would introduce
complexity. Therefore “state 1” should not be MUST.

Mit freundlichen Grüßen / Best regards

Achim Kraus

Bosch Software Innovations GmbH
Communications (INST/ESY1)
Stuttgarter Straße 130
71332 Waiblingen
GERMANY
www.bosch-si.de
www.blog.bosch-si.com<http://www.blog.bosch-si.com>

Tel. +49 711 811-58139
achim.kraus@bosch-si.com<mailto:achim.kraus@bosch-si.com>

Registered office: Berlin, Register court: Amtsgericht Charlottenburg, HRB 148411 B
Executives: Dr.-Ing. Rainer Kallenbach; Michael Hahn


Von: Hannes Tschofenig [mailto:notifications@github.com]
Gesendet: Dienstag, 20. September 2016 17:54
An: OpenMobileAlliance/OMA_LwM2M_for_Developers <OMA_LwM2M_for_Developers@noreply.github.com>
Cc: Kraus Achim (INST/ESY1) <Achim.Kraus@bosch-si.com>; Author <author@noreply.github.com>
Betreff: Re: [OpenMobileAlliance/OMA_LwM2M_for_Developers] OMA-DM-LightweightM2M-2016-0119-CR_firmware_update_state_machine (#136)


Based on discussions during today's conference call I have simplified the state machine drawing.

[statemachine-20sep2016]<https://cloud.githubusercontent.com/assets/1016643/18678076/2bba73f2-7f5b-11e6-955e-abd27054dd45.png>

**Comment by ThGarnier** [comment]
Now the Oct 16 TS 1.0 should anwser all questions related to Firmware Update please open new synthetic issue if still pending questions => To Be Closed

**Comment by boaks** [comment]
Fixed in OMA-TS-LightweightM2M-V1_0-20161007-D.
